src/emqx_sdv.erl

```diff
@@ -0,0 +1,179 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_sdv).
+
+%% for #message{} record
+%% no need for this include if we call emqx_message:to_map/1 to convert it to a map
+-include_lib("emqx_plugin_helper/include/emqx.hrl").
+
+%% for hook priority constants
+-include_lib("emqx_plugin_helper/include/emqx_hooks.hrl").
+
+%% for logging
+-include_lib("emqx_plugin_helper/include/logger.hrl").
+
+-include("emqx_sdv.hrl").
+
+-export([
+    hook/0,
+    unhook/0,
+    start_link/0
+]).
+
+-export([
+    on_config_changed/2,
+    on_health_check/1
+]).
+
+%% Hook callbacks
+-export([
+    on_message_publish/1,
+    on_delivery_completed/2
+]).
+
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2
+]).
+
+-define(SDV_FANOUT_DATA_TOPIC, "$SDV-FANOUT/data").
+-define(SDV_FANOUT_TRIGGER_TOPIC, "$SDV-FANOUT/trigger").
+
+%% @doc
+%% Called when the plugin application start
+hook() ->
+    %% Handle batch for fanout from SDV platform
+    %% Hook it with higher priority than retainer
+    %% which is also higher than rule engine
+    %% so the message can be terminated here at this plugin,
+    %% but not to leak to retainer or rule engine
+    emqx_hooks:add('message.publish', {?MODULE, on_message_publish, []}, ?HP_RETAINER + 1),
+    %% Handle PUBACK from subscribers (vehicles)
+    emqx_hooks:add('delivery.completed', {?MODULE, on_delivery_completed, []}, ?HP_HIGHEST),
+    ok.
+
+%% @doc
+%% Called when the plugin stops
+unhook() ->
+    emqx_hooks:del('message.publish', {?MODULE, on_message_publish}),
+    emqx_hooks:del('delivery.completed', {?MODULE, on_delivery_completed}),
+    ok.
+
+%%--------------------------------------------------------------------
+%% Hook callbacks
+%%--------------------------------------------------------------------
+
+%% @doc
+%% Called when a message is published by the SDV platform.
+on_message_publish(
+    #message{
+        timestamp = Ts, topic = <<?SDV_FANOUT_DATA_TOPIC, $/, DataID/binary>>, payload = Payload
+    } = Message
+) ->
+    Headers = Message#message.headers,
+    ok = emqx_sdv_fanout_data:insert_new(DataID, Payload, Ts),
+    {stop, Message#message{headers = Headers#{allow_publish => false}}};
+on_message_publish(#message{topic = <<?SDV_FANOUT_TRIGGER_TOPIC>>, payload = Payload} = Message) ->
+    Headers = Message#message.headers,
+    case emqx_sdv_fanout_dispatcher:trigger(Payload) of
+        ok ->
+            {stop, Message#message{headers = Headers#{allow_publish => false}}};
+        {error, Reason} ->
+            %% Disconnect the client, this seems to be the only way to notify the client
+            %% about the error, as EMQX message.publish hook does not support returning
+            %% PUBACK reason codes.
```

Comment:
We may put a header and detect it in `message.puback` hook.

Reply:
will keep disconnect for now
updated the comment.